cpp/segtree.hpp

```diff
@@ -0,0 +1,339 @@
+                sum = op(sum, data[l * 2]);
+                l = l * 2 + 1;
+            }
+        }
+        return l - sz;
+    }
+    /**
+     * @brief f(prod([l, r))) = trueとなる最小のlを返す
+     * 
+     * @tparam F
+     * @param r 半開区間の終端
+     * @param f 判定関数
+     * @return int 最小のl
+     */
+    template <typename F>
+    int min_left(int r, F f) const {
+        assert(f(e()));
+        if (r == 0) return 0;
+        r += sz - 1;
+        while (r % 2 == 1) r >>= 1;
+        S sum = e();
+        while(f(op(sum, data[r]))) {
+            if (__builtin_clz(r) != __builtin_clz(r-1)) return 0;
+            sum = op(sum, data[r]);
+            r--;
+            while (r % 2 == 1) r >>= 1;
+        }
+        while (r < sz) {
+            if (!f(op(data[r * 2 + 1], sum))) r = r * 2 + 1;
+            else {
+                sum = op(data[r * 2 + 1], sum);
+                r *= 2;
+            }
+        }
+        return r + 1 - sz;
+    }
+};
+
+/**
+ * @brief 積のファンクタが静的な場合のセグメント木の実装
+ * 
+ * @tparam S モノイドの型
+ * @tparam Op 積のファクタ
+ * @tparam E 積の単位元を返すファンクタ
+ */
+template <typename S, typename Op, typename E>
+class StaticSegTree : public SegTreeBase<S, StaticSegTree<S, Op, E>> {
+    using BaseType = SegTreeBase<S, StaticSegTree<S, Op, E>>;
+
+    inline static Op operator_object;
+    inline static E identity_object;
+public:
+    S op(const S& a, const S& b) const { return operator_object(a, b); }
+    S e() const { return identity_object(); }
+
+    /**
+     * @brief デフォルトコンストラクタ
+     * 
+    */
+    StaticSegTree() = default;
+    /**
+     * @brief コンストラクタ
+     * 
+     * @param n 要素数
+     */
+    explicit StaticSegTree(int n) : BaseType(n) {
+        this->construct_data();
+    }
+    /**
+     * @brief コンストラクタ
+     * 
+     * @param v 初期の要素
+     */
+    explicit StaticSegTree(const std::vector<S>& v) : StaticSegTree(v.size()) {
+        this->initialize(v);
+    }
+};
+
+/**
+ * @brief コンストラクタで関数オブジェクトを与えることで積を定義するセグメント木の実装
+ * テンプレート引数を省略することができ、ラムダ式が使える
+ * 
+ * @tparam S モノイドの型
+ * @tparam Op 積の関数オブジェクトの型
+ */
+template <typename S, typename Op>
+class SegTree : public SegTreeBase<S, SegTree<S, Op>> {
+    using BaseType = SegTreeBase<S, SegTree<S, Op>>;
+
+    Op operator_object;
+    S identity;
+
+public:
+    S op(const S& a, const S& b) const { return operator_object(a, b); }
+    S e() const { return identity; }
+
+    /**
+     * @brief デフォルトコンストラクタ
+    */
+    SegTree() = default;
+    /**
+     * @brief コンストラクタ
+     * 
+     * @param n 要素数
+     * @param op 積の関数オブジェクト
+     * @param identity 単位元
+     */
+    explicit SegTree(int n, Op op, const S& identity) : BaseType(n), operator_object(std::move(op)), identity(identity) {
+        this->construct_data();
+    }
+    /**
+     * @brief コンストラクタ
+     * 
+     * @param v 初期の要素
+     * @param op 積の関数オブジェクト
+     * @param identity 単位元
+     */
+    explicit SegTree(const std::vector<S>& v, Op op, const S& identity) : SegTree(v.size(), op, identity) {
```

Comment:
`op` をムーブするならムーブする、しないならしないで統一しましょう。

```suggestion
    explicit SegTree(int n, Op op, const S& identity) : BaseType(n), operator_object(std::move(op)), identity(identity) {
        this->construct_data();
    }
    /**
     * @brief コンストラクタ
     * 
     * @param v 初期の要素
     * @param op 積の関数オブジェクト
     * @param identity 単位元
     */
    explicit SegTree(const std::vector<S>& v, Op op, const S& identity) : SegTree(v.size(), std::move(op), identity) {
```